x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/preprocessing/Multi.java

```diff
@@ -0,0 +1,234 @@
+            throw new IllegalArgumentException("processors must be an array of objects with at least length 2");
+        }
+        this.custom = custom != null && custom;
+        Set<String> consumedOutputFields = new HashSet<>();
+        List<String> inputFields = new ArrayList<>(processors[0].inputFields());
+        Map<String, String> originatingOutputFields = new LinkedHashMap<>();
+        for (String outputField : processors[0].outputFields()) {
+            originatingOutputFields.put(outputField, processors[0].getOutputFieldType(outputField));
+        }
+        for (int i = 1; i < processors.length; i++) {
+            final PreProcessor processor = processors[i];
+            for (String inputField : processor.inputFields()) {
+                if (originatingOutputFields.containsKey(inputField) == false) {
+                    inputFields.add(inputField);
+                } else {
+                    consumedOutputFields.add(inputField);
+                }
+            }
+            for (String outputField : processor.outputFields()) {
+                originatingOutputFields.put(outputField, processor.getOutputFieldType(outputField));
+            }
+        }
+        Map<String, String> outputFields = new LinkedHashMap<>();
+        for (var outputField : originatingOutputFields.entrySet()) {
+            if (consumedOutputFields.contains(outputField.getKey()) == false) {
+                outputFields.put(outputField.getKey(), outputField.getValue());
+            }
+        }
+        this.outputFields = outputFields;
+        this.inputFields = inputFields.toArray(String[]::new);
+    }
+
+    public Multi(StreamInput in) throws IOException {
+        this.processors = in.readNamedWriteableList(PreProcessor.class).toArray(PreProcessor[]::new);
+        this.custom = in.readBoolean();
+        this.outputFields = in.readOrderedMap(StreamInput::readString, StreamInput::readString);
+        this.inputFields = in.readStringArray();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteableList(Arrays.asList(processors));
+        out.writeBoolean(custom);
+        out.writeMap(outputFields, StreamOutput::writeString, StreamOutput::writeString);
+        out.writeStringArray(inputFields);
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this);
+    }
+
+    @Override
+    public List<String> inputFields() {
+        return Arrays.asList(inputFields);
+    }
+
+    @Override
+    public List<String> outputFields() {
+        return new ArrayList<>(outputFields.keySet());
+    }
+
+    @Override
+    public void process(Map<String, Object> fields) {
+        for (PreProcessor processor : processors) {
+            processor.process(fields);
+        }
+    }
+
+    @Override
+    public Map<String, String> reverseLookup() {
+        // TODO
+        throw new UnsupportedOperationException();
```

Comment:
Is it fine to throw here? It seems to be used in `InferenceDefinition.getDecoderMap`.

Reply:
I don't know if we could ever support a reverseLookup in a multi-processor. Only if the input field count is one. If the input field count is more than one, we have no way of knowing which importances map back from the output fields to which input field. 

We should definitely have a better error thrown here